### PR TITLE
Create ProfanityCheckerJob.java

### PR DESCRIPTION
**Pull request by jrosales**

### CODE REVIEW

File: src/main/java/com/gitinspector/scheduling/ProfanityCheckerJob.java [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+
+    private String STATS_TRACKER_COMMIT_MSG_NAME = "reposWithProfanityInCommitMsg";
+
+    private GitHub gitHub;
+
+    private List<String> searchTerms;
+
+    private List<Pattern> searchPatterns;
+
+    @Value("${git-inspector.profanityListPath:''}")
```
Comment (bcorbett) [inline src/main/java/com/gitinspector/scheduling/ProfanityCheckerJob.java]: I'm thinking we might want to start by having this supplied in the constructor. We were avoiding having a spring dependency to minimize requirements for projects wanting to use us as a jar. If we do move it to the constructor, I'm not sure if it's best to supply a File handle to the profanity list or to just pass a List<String>. Thoughts?

File: src/main/java/com/gitinspector/scheduling/ProfanityCheckerJob.java [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+        GHBranch masterBranch = repository.getBranches().get(repository.getMasterBranch());
+        GHCommit masterTipCommit = repository.getCommit(masterBranch.getSHA1());
+        List<GHCommit> masterCommits = new ArrayList<>();
+        assembleCommits(repository, masterTipCommit, masterCommits, numberOfDaysThreshold);
+
+        for (GHCommit commit : masterCommits) {
+            hasProfanity = false;
+
+            for (String term : searchTerms) {
+                String commitMessage = commit.getCommitShortInfo().getMessage();
```
Comment (bcorbett) [inline src/main/java/com/gitinspector/scheduling/ProfanityCheckerJob.java]: Let's move this variable assignment to happen before we start looping over the searchTerms. No reason to keep re-assigning it.

File: src/main/java/com/gitinspector/scheduling/ProfanityCheckerJob.java [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+
+                Pattern pattern = Pattern.compile(term);
+                if (pattern.matcher(commitMessage).find()) {
+                    reportResult.addViolation(
+                        new BadCommit(getOrgNameFromRepoName(repoFullName), repoFullName, getOwnerUsername(repoFullName),
+                            commit.getCommitShortInfo().getCommitter().getName(), commit.getSHA1()));
+                    hasProfanity = true;
```
Comment (bcorbett) [inline src/main/java/com/gitinspector/scheduling/ProfanityCheckerJob.java]: Let's break here. No need to create more than 1 BadCommit violation per commit.

File: src/main/java/com/gitinspector/scheduling/ProfanityCheckerJob.java (hunk 1 above)
Comment (jrosales) [inline src/main/java/com/gitinspector/scheduling/ProfanityCheckerJob.java]: I like the list approach as it makes it a bit more flexible for folks to decide where the profanity list can be stored.

File: src/main/java/com/gitinspector/scheduling/ProfanityCheckerJob.java [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+    private String WITH_PROFANE_LANGUAGE_IN_COMMIT_MSG = "WithProfaneLanguageInCommitMsg";
+
+    private String STATS_TRACKER_SRC_NAME = "reposWithProfanityInSrc";
+
+    private String STATS_TRACKER_COMMIT_MSG_NAME = "reposWithProfanityInCommitMsg";
+
+    private GitHub gitHub;
+
+    private List<Pattern> searchPatterns;
```
Comment (bcorbett) [inline src/main/java/com/gitinspector/scheduling/ProfanityCheckerJob.java]: Unused variable

File: src/main/java/com/gitinspector/scheduling/ProfanityCheckerJob.java [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+        for (GHCommit commit : masterCommits) {
+            hasProfanity = false;
+            String commitMessage = commit.getCommitShortInfo().getMessage();
+
+            for (String term : profanityList) {
+                Pattern pattern = Pattern.compile(term);
+                if (pattern.matcher(commitMessage).find()) {
+                    reportResult.addViolation(
+                        new BadCommit(getOrgNameFromRepoName(repoFullName), repoFullName, getOwnerUsername(repoFullName),
```
Comment (bcorbett) [inline src/main/java/com/gitinspector/scheduling/ProfanityCheckerJob.java]: What do you think about logging the profane word that was encountered? If we want to do that we should need a new Violation class, say CommitWithProfanity which could extend BadCommit and add a property to represent the profane word. If we do that should we take it a step further than log all of the profane words too? If we did that we wouldn't want multiple violations, we'd want just want the CommitWithProfanity violation to store a List<String> for the words.

File: src/main/java/com/gitinspector/scheduling/ProfanityCheckerJob.java [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+        String repoFullName = repository.getFullName();
+
+        for (String term : profanityList) {
+            GHContentSearchBuilder search = gitHub.searchContent();
+            search.repo(repository.getFullName()).q("\"" + term + "\"");
+            PagedSearchIterable<GHContent> list = search.list();
+
+            if (list.getTotalCount() > 0) {
+                reportResult.addViolation(
+                    new Violation(getOrgNameFromRepoName(repoFullName), repoFullName, getOwnerUsername(repoFullName)));
```
Comment (bcorbett) [inline src/main/java/com/gitinspector/scheduling/ProfanityCheckerJob.java]: Same question here. We aren't tracking either the file or the profane word(s) we found. I think we need a new Violation, something like FileWithProfanity which might add "filePath" and "profaneWords" properties so they could be logged out.

File: src/main/java/com/gitinspector/scheduling/ProfanityCheckerJob.java (hunk 5 above)
Comment (jrosales) [inline src/main/java/com/gitinspector/scheduling/ProfanityCheckerJob.java]: Yeah, I prefer the latter. Let me whip something up so we can have the new violation.

File: src/main/java/com/gitinspector/scheduling/ProfanityCheckerJob.java (hunk 6 above)
Comment (jrosales) [inline src/main/java/com/gitinspector/scheduling/ProfanityCheckerJob.java]: I was hoping that I could get the LoC from the search results but that does not seem to be possible with the kohsuke API.

Comment (bcorbett) [inline src/main/java/com/gitinspector/scheduling/ProfanityCheckerJob.java]: That's fine. Can we at least get the path?

Comment (jrosales) [inline src/main/java/com/gitinspector/scheduling/ProfanityCheckerJob.java]: Yeah, updated the read.

File: src/main/java/com/gitinspector/scheduling/ProfanityCheckerJob.java [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+    private GitStatisticsTracker checkFiles(GHRepository repository, ReportResult<Violation,
+            StringStatistic> reportResult) {
+        GitStatisticsTracker srcTracker = new GitStatisticsTracker(STATS_TRACKER_SRC_NAME);
+        String repoFullName = repository.getFullName();
+
+        Map<String, List<String>> foundTermMap = new HashMap<>();
+
+        for (String term : profanityList) {
+            GHContentSearchBuilder search = gitHub.searchContent();
+            search.repo(repository.getFullName()).q("\"" + term + "\"");
```
Comment (bcorbett) [inline src/main/java/com/gitinspector/scheduling/ProfanityCheckerJob.java]: When we first test this rule, let's do it with only 1 or 2 terms. Searching for one term per repo could end up being really slow. I wonder if we could search for multiple terms at a time. No need to change it here. We can figure that out later.

**Comment by bcorbett** [comment]
Looks good!

File: src/main/java/com/gitinspector/scheduling/ProfanityCheckerJob.java (hunk 7 above)
Comment (jrosales) [inline src/main/java/com/gitinspector/scheduling/ProfanityCheckerJob.java]: :+1:

**Pull request merged**